app/javascript/mastodon/actions/compose.js

```diff
@@ -661,11 +662,11 @@ export function selectComposeSuggestion(position, token, suggestion, path) {
 
       dispatch(useEmoji(suggestion));
     } else if (suggestion.type === 'hashtag') {
-      completion    = `#${suggestion.name}`;
-      startPosition = position - 1;
+      completion    = suggestion.name.slice(token.length - 1);
+      startPosition = position + token.length;
     } else if (suggestion.type === 'account') {
-      completion    = getState().getIn(['accounts', suggestion.id, 'acct']);
-      startPosition = position;
+      completion    = `@${getState().getIn(['accounts', suggestion.id, 'acct'])}`;
+      startPosition = position - 1;
```

Comment:
Is there a reason for keeping the token in one case and replacing it in the other? If token length is variable, wouldn't an issue occur?

Reply:
Logic is as follows:

- For `@`, bring up autocomplete when full-width `＠` is typed, but replace with ASCII when a suggestion is selected. Since usernames are ASCII, it will look better with an ASCII `＠` (also I didn't check if the server will parse full-width at all)
- For hashtags, allow keeping the full-width `＃`, since it is part of the text, and hashtags are unicode, and full-width `＃` is handled server-side correctly. Also, this finally keeps the casing the user typed when a suggestion is selected.

Reply:
> Also, this finally keeps the casing the user typed when a suggestion is selected.

Wait, isn't that a regression? If someone starts typing in lowercase and selects the properly-capitalized hashtag, it won't be capitalized anymore. I thought we specifically had code to keep the “most capitalized one”.